### ISSUE TITLE
Dynamically Scheduled parallel_for (questions)

### ISSUE THREAD

**Issue by vblanco20-1** [cpp-taskflow/cpp-taskflow#31]
Hello.
At the moment, im moving one of my game experiments to taskflow (a spaceship battle simulation with 150.000 game objects moving around). [details](http://victor.madtriangles.com/code%20experiment/2018/06/11/post-ecs-battle-huge.html)  

For this, ive implemented taskflow to schedule my whole game loop, having each of the game simulation steps be a task. This has worked quite well, and it runs fine, but this is not at full speed yet. Just running multiple systems at once increases parallelism but not enough.

In the original implementation, i was using parallel for constantly (the c++17 std parallel algorithm version), and it was netting me a 6-7 times scalability on a ryzen with 8 cores and 16 threads. I tested running std::for_each inside tasks and it actually runs and scales, but i dont think thats good as i would like to leave taskflow to schedule everything instead.

On the engine, i have a "schedule" step that builds the task graph for the whole engine, and then i execute it by making the main thread wait until everything is finished. 

This means there is a delay beetween the task creation and the task execution, and this causes considerable issues with the parallel for. The parallel for creates a few tasks but there is no dynamic version. On the engine, i have a task that "prepares" the game objects that will run for a system (for example, the renderables), and then have a parallel for that goes over all that list of renderables and calculates culling. Next, a different system has a different task that just iterates over the visible renderables and draws them to the screen. Sadly there is no way to create tasks within tasks, wich would solve this instantly.

```cpp
//pseudocode mockup
//-- somewhere , basically  globals (non owning, temporal) so they can be shared from place to place
vector<RenderObject*> *renderables;// pointer to a renderables vector, can be any size from 10 to 200.000
CameraData * camdata;//also global

//There should be no data races by design given the constraints within tasks.


//--------------------------------
//over multiple systems and classes but simplified

//gameplay systems scheduling
auto root_task = task_engine.placeholder();

auto game_final_task = GameSystems.Schedule(root_task);

//grabs renderable data and camera data and stores them in the temporal globals to share them across time. 
auto prepare_cull_task = task_engine.silent_emplace([&]() {
camdata = Engine.GetCameraData();
renderables = Engine.GetRenderables(); //gather all the renderables and put them in an array, then point that to the global one so it can be used out of scope.

}

//performs culling
auto cull_task = task_engine.silent_emplace([&]() {
    //serial, should actually be parallel
    for(auto r : Renderables){
       r->cull(camdata);
    }
}

//renders the renderable list
auto render_task =  task_engine.silent_emplace([&]() {
   //serial becouse gpu
    for(auto r: Renderables)    {
         if(r->visible)  Engine.DrawRenderable(r);
    }
}


prepare_cull_task.gather(game_final_task);
cull_task.gather(prepare_cull_task);
render_task.gather(cull_task);

//execute the whole engine iteration
task_engine.wait_for_all();


```

The provided parallel_for will not work in a case like this, becouse it needs to have a set size. Even if i create the parallel for task on the schedule step, this will not work becouse the amount of renderables is set by the gameplay systems, wich can create or destroy game objects.  

For this, i was going to implement my own dynamically scheduled parallel for, more similar to how std::for_each(parallel) works, but i wanted to ask about the design. 

My plan is to have a new type of high level task creator that supports this pattern. It would have a lambda for the "preparation" step, and then it will create a customizable number of tasks. (for example 4 tasks). Those tasks will execute the work prepared on the preparation step.

For example, my culling system has the preparation step of grabbing all the renderables, and they are 100.000. Becouse its that many, the intermediate step (that prepares the parallel for) divides it in batches of 4048 (configurable) and inserts the chunks into a parallel queue (just begin/end ranges). Then the worker tasks just dequeue the work chunks and execute the function inside.

Is a pattern like that alright? Given  how common the pattern is, i would like to have it be generic and add it to the library.

**Comment by twhuang-uiuc** [comment]
@vblanco20-1  Thanks for bringing this up. Yes we completely agree with the need of enabling "intra-task parallelism". Say, a task can spawn multiple tasks inside its little "arena" and these tasks won't break the dependency outside their world. We are researching on the interface and hopefully can work it out soon.

Regarding your question, I am a bit confused. Are you asking whether the pseudo code your wrote using Taskflow is correct? Or are you asking about the use of `parallel_for`? Please let me know. It seems to me your workflow is similar to `Map-Reduce` where you have multiple iterations that spawn parallel tasks and synchronize at some points using `wait_for_all`. Please let me know and I can help.

**Comment by twhuang-uiuc** [comment]
@vblanco20-1 Yes. We are researching on an interface to dynamic task parallelism. So far, Taskflow is static and users need to construct the graph before executing them. And your pseudo-code looks perfect for me.

**Comment by bsviglo** [comment]
Hi guys, I'd like to add that this functionality with dynamic task parallelism is very important in many cases. So I'm really waiting for it.

**Comment by twhuang-uiuc** [comment]
Yes. We are working on it now and will have a release these two days. Please stay tuned.

**Comment by bsviglo** [comment]
Hi guys, I've seen your commit related to the dynamic task parallelism. Looks cool!
But I have a question. Looking at diagram below it looks like subflow_B is executed concurrently with task A. Is it true? It seems strange because B runs subflow_B and B depends on A and should be executed after A and subflow_B (as nested taskflow to B) in turns should be executed right after A (on the second diagram I make desired dependencies)
![image](https://user-images.githubusercontent.com/6697628/44716632-94682d00-aac3-11e8-89d8-9ce9241924b5.png) ![image](https://user-images.githubusercontent.com/6697628/44716771-ef9a1f80-aac3-11e8-9ade-f9d33ac5dadc.png)

**Comment by vblanco20-1** [comment]
Wow thats awesome that it got added. Im going to try to implement it to my experimental engine when i find the time to do so, should be a great benchmark test (im comparing to std::for_each and other parallel stl)

**Comment by twhuang-uiuc** [comment]
@vblanco20-1 Thanks for your interests. Will be happy if you would be able to contribute those examples to Cpp-Taskflow.

**Comment by twhuang-uiuc** [comment]
@bsviglo No. B1, B2, and B3 will be spawned during the execution of B - which means A already finished at this point because A precedes B. That is, B1, B2, and B3 will run after A. I will added this to [frequently asked question](https://github.com/cpp-taskflow/cpp-taskflow/blob/master/doc/faq.md).

**Comment by vblanco20-1** [comment]
Ive been able to check the system out, it works like a charm!!

Reaching essentially the same performance if not slightly better than the std::for_each parallel. Now the thing is how do i profile this to show exact data :D
